### PR TITLE
Allow combination of class and protocol reify.

### PR DESCRIPTION
**Pull request by borkdude**

### CODE REVIEW

**Comment by borkdude** [comment]
@GreshamDanielStephens Looks like a solid approach! 

Note to self: all reified objects will implement

- sci.impl.types.Reified interface (which then dispatches to "protocols" (non-interfaces)) 
- optionally more Java interfaces.

Cool stuff. Maybe you can fix the native and node tests and as you said, maybe it needs simplification, bit already looks quite good to me!

**Comment by GreshamDanielStephens** [comment]
Seems generally accepted to not run some tests in native when they require some evaluated context to be passed to the eval* method. That's the case for the test I added that needs to create a function.

I did however manually test that this works by adding my tests reify block into `default-reify` and then running the test string.
```
./sci "(defprotocol Protocol (custom [this])) (let [r (reify Object (toString [this] \"obj-str\") clojure.lang.IDeref (deref [this] \"ideref-deref\") Protocol (custom [this] \"protocol-custom\"))] [(str r) (deref r) (custom r)])"
["obj-str" "ideref-deref" "protocol-custom"]
```

which gives the expected output :tada: 

Still need to run this against babashka tests, and look at any simplifications we want.

**Comment by borkdude** [comment]
Cool! Can you merge with the master branch as well?

**Comment by GreshamDanielStephens** [comment]
I'm about 90% sure that I've run the babashka tests correctly against this.
The normal tests all pass.
The lib tests have 6 errors, seems to be the same with the master branch.
They are a bit scary as the seem related to some deref's and reify calls, one of them hits here:
https://github.com/http-kit/http-kit/blob/059deac93b1662077d52e1eb81bef9db8c89d746/src/org/httpkit/client.clj#L148
```
ERROR in (get-test) (/home/dstephens/git/babashka/test-resources/lib_tests/httpkit/client_test.clj:17)
expected: (= 200 (:status (clojure.core/deref (client/get "https://postman-echo.com/get"))))
  actual: java.lang.NullPointerException: null
 at clojure.lang.Numbers.ops (Numbers.java:1068)
    clojure.lang.Numbers.gte (Numbers.java:263)
    clojure.lang.Numbers.gte (Numbers.java:3981)
    org.httpkit.sni_client$ssl_configurer.invokeStatic (sni_client.clj:44)
    org.httpkit.sni_client$ssl_configurer.invoke (sni_client.clj:37)
    org.httpkit.sni_client$ssl_configurer.invokeStatic (sni_client.clj:41)
    org.httpkit.sni_client$ssl_configurer.invoke (sni_client.clj:37)
    org.httpkit.client$make_client$reify__21181.configure (client.clj:148)
    ...
```

I will try and double check again later.

**Comment by borkdude** [comment]
@GreshamDanielStephens I will push this branch to a sci branch and also make a bb branch that depends on this sci branch and will run that in CI.

**Comment by borkdude** [comment]
Babashka CI running here:

https://github.com/babashka/babashka/tree/GreshamDanielStephens-reify

**Comment by borkdude** [comment]
Tests seem to pass. Your problem might be related to some dirty `target` folder or something. I added `lein do clean, test` to the `script/test` script now.

File: src/sci/impl/types.cljc [hunk 1]
```diff
@@ -10,16 +10,17 @@
...
+(deftype Reified [interfaces meths]
   IReified
-  (getInterface [_] interface)
+  (getInterfaces [_] interfaces)
   (getMethods [_] meths))
 
 (defn type-impl [x & _xs]
-  (or (when (instance? sci.impl.types.Reified x)
+  (or (when #?(:clj (instance? sci.impl.types.IReified x)
+               :cljs (clojure.core/satisfies? IReified x))
```
Comment (borkdude) [inline src/sci/impl/types.cljc:23]: @GreshamDanielStephens Why was this change needed?

File: src/sci/impl/reify.cljc [hunk 2]
```diff
@@ -2,34 +2,38 @@
...
-              ;; So far we only supported reify-ing one protocol at a time. This
-              ;; should be addressed in a future version
-              (let [[interface methods] (first classes->methods)]
-                (t/->Reified interface methods))))
-     :cljs (let [[interface methods] (first classes->methods)]
-             (t/->Reified interface methods))))
+                 :cljs _ctx) classes methods]
+  #?(:clj (let [{classes true protocols false} (group-by class? classes)
+
+                classes (if (seq protocols) (distinct (conj classes IReified)) classes)
```
Comment (borkdude) [inline src/sci/impl/reify.cljc:20]: I think you can pull `(seq protocols)` to a variable `protocols?` to avoid doing on again on line 27

File: src/sci/impl/reify.cljc [hunk 3]
```diff
@@ -2,34 +2,38 @@
...
+                reify-methods {'getInterfaces (constantly protocols)
+                               'getMethods (constantly methods)}
+                methods (if (seq protocols) (merge reify-methods methods) methods)
+
+                ;; The reify factories get the fn by classname and method name
+                ;; however it shouldn't actually matter which class the method
+                ;; belongs to, as different classes may have the same method
+                ;; names with different arities.
+                ;; classes->methods puts all methods for all classes.
+                classes->methods (into {} (map vector class-names (repeat methods)))]
```
Comment (borkdude) [inline src/sci/impl/reify.cljc:34]: Is there a specific reason you mashed everything together instead of leaving it structured?

Comment (borkdude) [reply src/sci/impl/reify.cljc:34]: My concern here is that two classes could have the same method name.

**Comment by borkdude** [comment]
Lumping all method together is probably not a good idea. Here is an example of where that might go wrong:

```
user=> (definterface Foo (foo [_]))
user.Foo
user=> (definterface Bar (foo [_ _]))
user.Bar
user=> (def x (reify Foo (foo [_ _] :foo) Bar (foo [_ _ _] :bar)))
#'user/x
user=> (.foo x 1)
:foo
user=> (.foo x 1 2)
:bar 
```

File: src/sci/impl/types.cljc (hunk 1 above)
Comment (GreshamDanielStephens) [reply src/sci/impl/types.cljc:23]: to be honest this confused me, and I'm pretty open to a better solution.
This stems from lack of clojureScript understanding I expect.
I had a test failure when using just `instance?`, I added a println for debugging and this is what I got.
```
(println x (instance? Reified x) (instance? IReified x) (satisfies? IReified x))

#object[sci.impl.types.Reified] true false true
```
So here I see in clojureScript I have an object of type Reified, which is an instance of Reified, but not of it's protocol IReified.
I started searching how to check it implements the protocol and found `satisfies?`
I assume this is a difference in how javascript and java work, and to do with IReified being an interface.

Testing again, satisfies looks like it works for both so I can go with that.

Comment (borkdude) [reply src/sci/impl/types.cljc:23]: In the spirit of only making changes that are necessary, why don't leave it like it was?

`(instance? sci.impl.types.Reified x)` for both cases?

`satisfies?` is a slow operation in Clojure(Script)

Comment (GreshamDanielStephens) [reply src/sci/impl/types.cljc]: `(instance? sci.impl.types.Reified x)` won't work for the new clj implementation of reify, since it directly implements IReified again without touching Reified. Didn't know satisfies is slow, I guess I can wrap just the class bit in the case `(instance? #?(:clj sci.impl.types.IReified :clj sci.impl.types.Reified) x)`?

File: src/sci/impl/reify.cljc (hunk 3 above)
Comment (GreshamDanielStephens) [reply src/sci/impl/reify.cljc:35]: Added the tests that we talked about in slack

File: src/sci/impl/types.cljc (hunk 1 above)
Comment (borkdude) [reply src/sci/impl/types.cljc]: OK, let's do that (the second conditional should be `:cljs`)

**Comment by borkdude** [comment]
Merged via other PR.
